src/dos/program_keyb.cpp

```diff
@@ -20,112 +20,385 @@
-		                               tried_cp,
-		                               cp_filename.c_str());
+		for (auto& entry : table) {
+			entry.first.resize(max_length, ' ');
+		}
 	}
 
-	// Switch if loading the layout succeeded
-	if (rcode == KEYB_NOERROR) {
-		rcode = DOS_SwitchKeyboardLayout(temp_line.c_str(), tried_cp);
+	const auto margin = std::string(target_len, ' ');
+
+	bool first_row = true;
+	for (auto& entry : table) {
+		if (first_row) {
+			message += script_msg;
+			first_row = false;
+		} else {
+			message += margin;
+		}
+
+		message += entry.first;
+		if (show_shortcuts) {
+			message += std::string(LargeSpacingSize, ' ');
+			message += AnsiYellow + entry.second + AnsiReset;
+		}
+		message += "\n";
+	}
+
+	const auto warn = DOS_GetCodePageWarning(dos.loaded_codepage);
+	if (warn) {
+		message += "\n";
+		message += format_str(MSG_Get("PROGRAM_KEYB_WARNING_CODE_PAGE"),
+		                      dos.loaded_codepage);
+		message += "\n";
+		switch (*warn) {
+		case CodePageWarning::DottedI:
+			message += MSG_Get("PROGRAM_KEYB_WARNING_DOTTED_I");
+			break;
+		default: assert(false); break;
+		}
+		message += "\n";
 	}
 
-	log_keyboard_code(rcode, temp_line, tried_cp);
+	print_message();
 }
 
-void KEYB::AddMessages() {
-	MSG_Add("PROGRAM_KEYB_INFO","Codepage %i has been loaded.\n");
-	MSG_Add("PROGRAM_KEYB_INFO_LAYOUT","Codepage %i has been loaded for layout %s.\n");
+void KEYB::AddMessages()
+{
 	MSG_Add("PROGRAM_KEYB_HELP_LONG",
-	        "Configure a keyboard for a specific language.\n"
+	        "Configure a keyboard layout and screen font.\n"
 	        "\n"
 	        "Usage:\n"
-	        "  [color=light-green]keyb[reset] [color=light-cyan][LANGCODE][reset]\n"
-	        "  [color=light-green]keyb[reset] [color=light-cyan]LANGCODE[reset] [color=white]CODEPAGE[reset] [color=white][CODEPAGEFILE][reset]\n"
+	        "  [color=light-green]keyb[reset]\n"
+	        "  [color=light-green]keyb[reset] /list\n"
+	        "  [color=light-green]keyb[reset] [color=light-cyan]LAYOUT[reset] [[color=white]CODEPAGE[reset]] /rom\n"
+	        "  [color=light-green]keyb[reset] [color=light-cyan]LAYOUT[reset] [[color=white]CODEPAGE[reset] [[color=white]CPIFILE[reset]]]\n"
 	        "\n"
 	        "Parameters:\n"
-	        "  [color=light-cyan]LANGCODE[reset]      language code or keyboard layout ID\n"
-	        "  [color=white]CODEPAGE[reset]      code page number, such as [color=white]437[reset] and [color=white]850[reset]\n"
-	        "  [color=white]CODEPAGEFILE[reset]  file containing information for a code page\n"
+	        "  [color=light-cyan]LAYOUT[reset]    keyboard layout code\n"
+	        "  [color=white]CODEPAGE[reset]  code page number, such as [color=white]437[reset] or [color=white]850[reset]\n"
+	        "  [color=white]CPIFILE[reset]   screen font file, in CPI or CPX format\n"
+	        "  /list     displays available keyboard layout codes\n"
+	        "  /rom      if possible, uses screen font from display adapter ROM\n"
```

Comment:
Command form is better:

"display available keyboard layout codes"
"use screen font from display adapter ROM if possible"

etc.

Reply:
OK, changed.